src/lib/Microsoft.Health.Events/EventCheckpointing/StorageCheckpointClient.cs

```diff
@@ -0,0 +1,240 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Timers;
+using Azure;
+using Azure.Storage.Blobs;
+using Azure.Storage.Blobs.Models;
+using EnsureThat;
+using Microsoft.Health.Events.Model;
+
+namespace Microsoft.Health.Events.EventCheckpointing
+{
+    public class StorageCheckpointClient : ICheckpointClient
+    {
+        private ConcurrentDictionary<string, Checkpoint> _checkpoints;
+        private BlobContainerClient _storageClient;
+        private static System.Timers.Timer _publisherTimer;
+        private int _publishTimerInterval = 10000;
+
+        public StorageCheckpointClient(StorageCheckpointOptions options)
+        {
+            EnsureArg.IsNotNull(options);
+            EnsureArg.IsNotNullOrWhiteSpace(options.BlobPrefix);
+            EnsureArg.IsNotNullOrWhiteSpace(options.BlobStorageConnectionString);
+            EnsureArg.IsNotNullOrWhiteSpace(options.BlobContainerName);
+
+            BlobPrefix = options.BlobPrefix;
+
+            _checkpoints = new ConcurrentDictionary<string, Checkpoint>();
+            _storageClient = new BlobContainerClient(options.BlobStorageConnectionString, options.BlobContainerName);
+
+            SetPublisherTimer();
+        }
+
+        public string BlobPrefix { get; }
+
+        public async Task UpdateCheckpointAsync(Checkpoint checkpoint)
+        {
+            EnsureArg.IsNotNull(checkpoint);
+            EnsureArg.IsNotNullOrWhiteSpace(checkpoint.Id);
+
+            var blobName = $"{BlobPrefix}/checkpoint/{checkpoint.Id}";
+            var blobClient = _storageClient.GetBlobClient(blobName);
+
+            var metadata = new Dictionary<string, string>()
+            {
+                { "LastProcessed", checkpoint.LastProcessed.DateTime.ToString("MM/dd/yyyy hh:mm:ss.fff tt") },
+            };
+
+            try
+            {
+                try
+                {
+                    await blobClient.SetMetadataAsync(metadata);
+                }
+                catch (RequestFailedException ex) when ((ex.ErrorCode == BlobErrorCode.BlobNotFound) || (ex.ErrorCode == BlobErrorCode.ContainerNotFound))
+                {
+                    using (var blobContent = new MemoryStream(Array.Empty<byte>()))
+                    {
+                        await blobClient.UploadAsync(blobContent, metadata: metadata).ConfigureAwait(false);
```

Comment:
Will this command also create the container if its not present?

Reply:
It will not create a container. I was thinking we should validate that the container exists on startup or something like that, like you suggested in another comment. 

We will have an arm template create the container, so in theory the container should exist unless someone goes and deletes it.